src/main/java/org/mockito/internal/framework/DefaultMockitoSession.java

```diff
@@ -0,0 +1,44 @@
+package org.mockito.internal.framework;
+
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.mockito.MockitoSession;
+import org.mockito.exceptions.misusing.RedundantListenerException;
+import org.mockito.internal.exceptions.Reporter;
+import org.mockito.internal.junit.TestFinishedEvent;
+import org.mockito.internal.junit.UniversalTestListener;
+import org.mockito.internal.util.MockitoLogger;
+import org.mockito.quality.Strictness;
+
+public class DefaultMockitoSession implements MockitoSession {
+
+    private final Object testClassInstance;
+    private final UniversalTestListener listener;
+
+    public DefaultMockitoSession(Object testClassInstance, Strictness strictness, MockitoLogger logger) {
+        this.testClassInstance = testClassInstance;
+        listener = new UniversalTestListener(strictness, logger);
+        try {
+            Mockito.framework().addListener(listener);
```

Comment:
> Why add a listener here?

When the session is started, the listener is added so that it can track creation of mocks and stubbing misuse.

In finish mocking we need to clean up state hence we remove the listener and report and findings.

Does it explain the workflow sufficiently? Do you suggest changes to the code / adding code comments?